resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/hadoopsteps/HadoopKerberosKeytabResolverStep.scala

```
@@ -0,0 +1,171 @@
+      } yield {
+        // Not necessary with [Spark-16742]
+        // Reliant on [Spark-20328] for changing to YARN principal
+        submissionSparkConf.set("spark.yarn.principal", principal)
+        submissionSparkConf.set("spark.yarn.keytab", keytab.toURI.toString)
+        logDebug("Logged into KDC with keytab using Job User UGI")
+        hadoopUGI.loginUserFromKeytabAndReturnUGI(
+          principal,
+          keytab.toURI.toString)
+      }
+    // In the case that keytab is not specified we will read from Local Ticket Cache
+    val jobUserUGI = maybeJobUserUGI.getOrElse(hadoopUGI.getCurrentUser)
+    // It is necessary to run as jobUserUGI because logged in user != Current User
+    jobUserUGI.doAs(new PrivilegedExceptionAction[Void] {
+      override def run(): Void = {
+        logDebug(s"Retrieved Job User UGI: $jobUserUGI")
+        originalCredentials = jobUserUGI.getCredentials
+        logDebug(s"Original tokens: ${originalCredentials.toString}")
+        logDebug(s"All tokens: ${originalCredentials.getAllTokens}")
+        logDebug(s"All secret keys: ${originalCredentials.getAllSecretKeys}")
+        dfs = FileSystem.get(hadoopConf)
+        // This is not necessary with [Spark-20328] since we would be using
+        // Spark core providers to handle delegation token renewal
+        renewer = jobUserUGI.getShortUserName
+        logDebug(s"Renewer is: $renewer")
+        credentials = new Credentials(originalCredentials)
+        dfs.addDelegationTokens(renewer, credentials)
+        // This is difficult to Mock and will require refactoring
+        tokens = credentials.getAllTokens.asScala
+        logDebug(s"Tokens: ${credentials.toString}")
+        logDebug(s"All tokens: ${tokens.mkString(",")}")
+        logDebug(s"All secret keys: ${credentials.getAllSecretKeys}")
+        null
+      }})
+    credentials.getAllTokens.asScala.isEmpty
+    tokens.isEmpty
+    if (tokens.isEmpty) logError("Did not obtain any Delegation Tokens")
```

I think we should throw an exception. Otherwise, it's hard to debug this downstream.